orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py

```diff
@@ -627,6 +627,44 @@ def run_step(model, input):
         _test_helpers.assert_values_are_close(ort_prediction, pt_prediction)
         _test_helpers.assert_gradients_match_and_reset_gradient(ort_model, pt_model, atol=1e-5)
 
+# This model with dynamic ignore_index requires torch version from commit 645119eaefd0dcf1afc54e9ad58678b5245dea78.
```

Comment:
@iK1D Does this mean that the fix also depends on a new version of PyTorch? If so, is there any way to backport it to work with PyTorch 1.8.x?

Reply:
I am not quite farmiliar with exporter. From this commit: https://github.com/pytorch/pytorch/commit/645119eaefd0dcf1afc54e9ad58678b5245dea78#diff-a915656e4a7a07553917af566ed0f559610872d9025063d82dacb9f5c37974c9, it said to lowering NLLLoss/CrossEntropyLoss to ATen code. I think without this, we cannot override the exporter to export the CrossEntropyLoss to ATenOp, then maybe we need exporter team to change their C++ code to support that?

Reply:
If I'm reading the diff correctly, it looks like previous behavior was for `cross_entropy_loss()` to be a Python wrapper around some code that dispatches to `torch._C._nn.nll_loss()` or `torch._C._nn.nll_loss2d()`. After that commit, it dispatches directly to `torch._C._nn.cross_entropy_loss()`, which gave us the new ability to register a symbolic for it.

I *think* the way to make this work with the old version of PyTorch would be to register a similar `NegativeLogLikelihoodLossInternal` op that supports dynamic `ignore_index`, and this would give us the ability to work with PyTorch 1.8.x (and support models using NLLLoss in the future). Does that make sense?

Reply:
Thanks @mrry It makes sense. The tricky part is, there are C++ code in exporter to fuse LogSoftmax->NegativeLogLikelihoodLoss to single SoftmaxCrossEntropyLoss. When we use NegativeLogLikelihoodLossInternal, this fuse will not work, so we also need a new transformer in our ORT side to do this fusion to reuse the kernel implementation of SoftmaxCrossEntropyLossInternal.

Reply:
Ah, that's a pity! Thanks for the explanation though.

Since it looks like a 1.9 release candidate branch is being prepared for PyTorch, we can probably just wait for the new version to be released, rather than adding a temporary new transformer in ORT.

Reply:
I just didn't see your comment before I pushed this new commit to support torch 1.8.1... I think it's good that we can support this torch version, what if there are users working on this old version...

Reply:
Oh wow, thanks for doing that 0 I didn't expect you to go to the trouble, but I appreciate it! Agreed that supporting 1.8.1 is worthwhile :).